DataFormats/Provenance/src/BranchType.cc

```diff
@@ -16,52 +20,37 @@ namespace edm {
     std::string const branchEntryInfo = "BranchEntryInfo";  // backward compatibility
     std::string const productProvenance = "ProductProvenance";
 
-    // Prefixes
-    std::string const run = "Run";
-    std::string const lumi = "LuminosityBlock";
-    std::string const event = "Event";
-
-    // Trees, branches, indices
-    std::string const runs = run + 's';
-    std::string const lumis = lumi + 's';
-    std::string const events = event + 's';
+    std::array<std::string, NumBranchTypes> const branchTypeNames{{"Event", "LuminosityBlock", "Run", "ProcessBlock"}};
+    std::size_t constexpr eventLumiRunSize = 3;
+    using NameArray = const std::array<std::string, eventLumiRunSize>;
 
-    std::string const runMeta = run + metaData;
-    std::string const lumiMeta = lumi + metaData;
-    std::string const eventMeta = event + metaData;
+    NameArray makeNameArray(std::string const& postfix) {
+      static_assert(InEvent == 0);
+      static_assert(InLumi == 1);
+      static_assert(InRun == 2);
```

Comment:
Now at least it won't compile if the values are changed. On the other hand, a loop would work automatically with value changes.

(this discussion makes me wonder if we have implicit assumptions on their numerical values somewhere else)

Reply:
In the version of this file in the current master release, there is no assumption about the enum values. In other files, especially in the IOPool subsystem, such assumptions already exist in the code (and one might find them other places if one looked).

When converting to std::array and using the BranchType as an index, I created the dependence. First here:
```
std::array<std::string, NumBranchTypes> const branchTypeNames{{"Event", "LuminosityBlock", "Run", "ProcessBlock"}};
```
Personally, I think the current version in the PR is good enough. It is simple and easy to read. I doubt the enum values will ever be reordered. If you want I could change initialization of the above array so that the enum values could  be reordered but still restricted to 0, 1, 2. If we want to support arbitrary values like InEvent = 100, InLumi = 101, and InRun = 102, then I would need to use something more like maps or add an additional layer of indirection. Note that there is also a presumption in the file that InProcess is 3. Most of the arrays have size 3 (run/lumi/event and no processBlock). Another alternative would be to remove the std::array's and go back to old way of implementing this.